README.md

```diff
@@ -107,46 +112,72 @@ The general idea of portals is summed up above: inset pre-rendering, activation,
 
-* The URL of the activated portal and the referring URL are available to normal navigations to the same extent they're available to a portal. Solutions to link decoration will apply to both.
+The below subsections explore the implementation of these restrictions in more detail.
 
+#### Storage access blocking
 
-TODO:
+Portaled pages that are cross-origin to their host will have no access to storage, similar to how an opaque-origin `<iframe>` behaves. (See this [discussion on the spec mechanism](https://github.com/whatwg/storage/issues/18#issuecomment-615336554).)
+
+We could attempt to address the threat by providing partitioned or ephemeral storage access, but then it is unclear how to transition to _unpartitioned_ storage upon activation. It would likely require some kind of web-developer-written merging logic. Completely blocking storage access is thus deemed simpler; portaled pages should not be doing anything which requires persistent storage before activation.
+
+This means that most existing content will appear "broken" when portaled by a cross-origin host. This necessitates an explicit opt-in to allow cross-origin content to be portaled, [discussed below](#opt-in). Such content might optionally "upgrade" itself to a credentialed view upon activation, as shown in the [example](#example) above.
+
+For a more concrete example, consider `https://aggregator.example/` which wants to prerender this GitHub repository using a `<portal>`. To make this work, GitHub would need to add the opt-in to allow the page to be portaled. Additionally, GitHub should add code to adapt their UI to show the logged-in view upon activation, by removing the "Join GitHub today" banner, and retrieving the user's credentials from storage and using them to replace the signed-out header with the signed-in header. Without such adapter code, activating the portal would show the user a logged-out view of GitHub in the top-level tab that the portal has been activated into. This would be a bad and confusing user experience, since the user is logged in to GitHub in all of their other top-level tabs.
+
+#### Communications channels that are blocked
+
+- [`postMessage()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) isn't allowed between a cross-origin portal and its container.
+- Blocked by preventing fetches within portals from using credentials until they're activated:
+  - The sequence of portal loads: The source page can encode its user ID into the order in which a sequence of URLs are loaded into portals. To prevent the target from correlating this ID with its own user ID without a navigation, a document loaded into a cross-origin portal is fetched without credentials and doesn't have access to either first-party storage or multi-keyed storage. It has to use [`requestStorageAccess()`](https://developer.mozilla.org/en-US/docs/Web/API/Document/requestStorageAccess) (or another TBD mechanism) to get access when it's activated.
```

Comment:
> The source page can encode its user ID into the order in which a sequence of URLs are loaded into portals

I don't understand this...is the plural "portals" suggesting that there are multiple `<portal>` elements and the order of their navigations can be used to signal a value to the target server? Or is it a sequence of navigations on a single `<portal>` to different URLs, the order of which is used to communicate the value.

Either case raises the question: why use a sequence of URLs rather than just putting the user id in the URL? e.g. 

`portal.src = 'https://tracker.com/user/bokand';

Reply:
> To prevent the target from correlating this ID with its own user ID without a navigation

What navigation is  "without a navigation" referring to? Navigating to a different URL in the portal? Or activating the portal? Either way, don't we want to prevent correlating IDs across sites in both these cases?

Reply:
I think the following answers are correct, but I am not sure:

> I don't understand this...is the plural "portals" suggesting that there are multiple `<portal>` elements and the order of their navigations can be used to signal a value to the target server? Or is it a sequence of navigations on a single `<portal>` to different URLs, the order of which is used to communicate the value.

Either would work.

> Either case raises the question: why use a sequence of URLs rather than just putting the user id in the URL? e.g.

I don't _think_ this case ("link decoration") is explicitly addressed yet. @jyasskin, do you agree? We talk about link decoration in the activation case, where activating is similar to clicking a link. But just the pure embedding case, with no activation, doesn't seem to be called out explicitly.

> What navigation is "without a navigation" referring to? Navigating to a different URL in the portal? Or activating the portal?

Activating the portal.

Once you activate the portal, you end up in the "Communications channels that match navigation" case; activating a portal should leak no more information than clicking a link does. The idea is that both are highly user-visible actions.

Reply:
> I don't think this case ("link decoration") is explicitly addressed yet. @jyasskin, do you agree?

In particular, currently we have

> **The URL of the activated portal** and the referring URL are available to portals to the same extent they're available to normal navigations. Solutions to link decoration will apply to both.

but we don't talk about the URL of the portal pre-activation.

Reply:
Without activation the other site can't access its own credentials, which means this is more similar embedding a subresource such as an image (which in the future won't have first-party cookies). With activation it's the general link decoration problem (with work going on in other groups to address that).

Reply:
Ah, right. With no activation ever, there's no threat. With eventual activation, then there's the threat of

> If we allowed communication, then the portaled content could be given the user ID from the host site. Then, after activation gives the portaled page access to first-party storage, it would join that user ID with information from its own first-party storage to perform cross-site tracking.

i.e. this opens us up to cross-site tracking in the same way that `<a>` does.

Reply:
> I don't think this case ("link decoration") is explicitly addressed yet.

Ah, thanks - I'd seen a few references to "link decoration" but wasn't certain what that referred to :)

> i.e. this opens us up to cross-site tracking in the same way that <a> does.

Got it - clicking on the portal/link implies the user has some relationship/trust with the destination so that's less of a problem than pre-activation.

Reply:
I spelled this out a bit more:

> Note that since a non-activated portal has no storage access, it cannot join any information stored in the URL with any of the portaled site's data. So it's only activation, which gives full first-party storage access, which creates a navigation-equivalent communications channel. This equivalence makes sense, as activating a portal is much like clicking a link.